process/block/gasConsumption.go

```diff
@@ -0,0 +1,529 @@
+	}
+	gasConsumedByOutgoingTransactions := gc.getTotalOutgoingGas()
+	txsLimitReachedForTotalOutgoing := gasConsumedByOutgoingTransactions+gasConsumedInSenderShard > bandwidthForOutgoingTransactions
+	txsLimitReachedInAnyShard := txsLimitReachedForSelfShard || txsLimitReachedForDestShard || txsLimitReachedForTotalOutgoing
+	if !txsLimitReachedInAnyShard {
+		// limit not reached, continue
+		// simply increment the lastTransactionIndex as this method might be called either from
+		// handling all transactions, either from handling pending, where the pending ones
+		// should have continuos indexes after the ones already included
+		// also add the consumed gas, for receiver shard too if needed
+		gc.lastTransactionIndex++
+		gc.totalGasConsumed[outgoingSelfKey] += gasConsumedInSenderShard
+		if isCrossShard {
+			gc.totalGasConsumed[outgoingDestKey] += gasConsumedInReceiverShard
+		}
+
+		return false, false, nil
+	}
+
+	// if limit is reached for any shard, transaction won't be added
+	// if mini blocks not handled yet:
+	//	- save the rest of transactions as pending
+	// if mini blocks handled and:
+	//	- there is some space left, add more transactions (handled above)
+	//	- there is no space left, return the last index added, do not save the pending transactions
+	if !gc.isMiniBlockSelectionDone {
+		// mini blocks not handled yet, save the rest of transactions, return no error
+		return true, true, nil
+	}
+
+	return false, true, nil
+}
+
+func (gc *gasConsumption) checkPendingOutgoingTransactions() error {
+	// checking if any pending transactions are left in order to fill the block
+	hasPendingTransactions := gc.isTransactionSelectionDone && len(gc.pendingTransactions) > 0
+	if !hasPendingTransactions {
+		return nil
+	}
+
+	for i := 0; i < len(gc.pendingTransactions); i++ {
+		tx := gc.pendingTransactions[i]
+		_, shouldStop, err := gc.checkOutgoingTransaction(tx)
+		if err != nil || shouldStop {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// TotalGasConsumed returns the total gas consumed for both incoming and outgoing transactions
+func (gc *gasConsumption) TotalGasConsumed() uint64 {
+	gc.mut.RLock()
+	defer gc.mut.RUnlock()
+
+	totalGasConsumed := uint64(0)
+	for _, gasConsumed := range gc.totalGasConsumed {
+		totalGasConsumed += gasConsumed
+	}
+
+	return totalGasConsumed
+}
+
+// GetLastMiniBlockIndexIncluded returns the last mini block index added
+func (gc *gasConsumption) GetLastMiniBlockIndexIncluded() int {
+	gc.mut.RLock()
+	defer gc.mut.RUnlock()
+
+	return gc.lastMiniBlockIndex
+}
+
+// GetLastTransactionIndexIncluded returns the last transactions index added
+func (gc *gasConsumption) GetLastTransactionIndexIncluded() int {
+	gc.mut.RLock()
+	defer gc.mut.RUnlock()
+
+	return gc.lastTransactionIndex
+}
+
+// DecreaseMiniBlockLimit reduces the mini block gas limit by a configured percentage
+func (gc *gasConsumption) DecreaseMiniBlockLimit() {
```

Comment:
I am not sure if we need to decrease the miniBlock limit.
Maybe we can discuss this.

Reply:
removed the methods